### PR TITLE
Netatmo: automatic search of the refresh interval

### LINKED ISSUE #3786: [Netatmo] Adjust polling according to upload interval
Data from the Netatmo add-on is currently only read every 10 minutes.
My mobile is buzzing with a notification that OpenHAB does not yet know, because only every second record is taken over.

The upload interval from the weather station to the cloud is currently 5 minutes.

Preferred Solution:
Using the timestamp, adding 5 minutes and 10 seconds and you will have the time when to poll the API next time.

See community: https://community.openhab.org/t/netatmo-refresh-interval/47100/12

### PR DESCRIPTION
**Pull request by lolodomo**
Fixes #3786

Signed-off-by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by cweitkamp** [comment]
Is there any other confirmation beside the comment in the community forum that the limit has been reduced? It has not yet been changed in the [developers guidelines](https://dev.netatmo.com/en-US/resources/technical/guides/ratelimits) by Netatmo. I am afraid I currently cannot confirm it for my own Netatmo devices / app. Do we need a new firmware for it? Is it a local change?

Personally I would prefer a more flexible solution to determine the polling interval like we have in the other modules. Instead of using a fixed property in the Thing we could switch to a config parameter `refreshInterval` with 5min as default value. The binding is already prepared for it. Wdyt?

https://github.com/openhab/openhab2-addons/blob/8452b59c97e5b9782b284d366e10d5754cf2bf28/addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/handler/NetatmoDeviceHandler.java#L207-L221

**Comment by JHCD** [comment]
I can confirm that the values (graphics) are displayed on the Netatmo home page in a 5 minute grid. They also differ and are not just "duplicated".

**Comment by lolodomo** [comment]
A setting is more flexible but add comfision especially when one particilar value is expected. I think this was the reason the setting was suppressed. In this case, this is the developer and not the final user who has to find the oprimal value.
My opinion is that the property is the good choice in this case.

**Comment by cweitkamp** [comment]
@lolodomo Why is the value expected? And why is this not given by defining the default value? It is the minimum restriction of Netatmo for querying their API. This could be handled easily by setting the min value of the `refreshInterval` to 5min.

@JHCD I agree and confirm it too. I have found the XLS/CSV export feature on the Neatmo website. The values for my devices are recorded nearly every 5min. But my Android App does not make use of this yet. It updates every 10min.

**Comment by lolodomo** [comment]
I agree, min+default values set to 5 minutes for the setting would be ok.
If everyone wants this solution rather mine, I am ok.

**Comment by lolodomo** [comment]
According to my current tests, `device.getLastStatusStore()` for the weather station is updated only every 10 minutes, not 5 minutes !

Like you requested, I added a setting rather than the property. But if we let the user define a smaller value than the real value like 5 minutes rather 10 minutes, the result is a little strange, the binding will update the data every 30 seconds in the frame between 5 and 10 minutes.

In fact, my feeling is that the binding should be able to identify the right refresh interval to be used after few minutes just by monitoring the difference between two different values of `device.getLastStatusStore()`.

So my proposal is to set internally (or through the property) an initial refresh interval to 5 minutes. As soon as the binding detects a different value for `device.getLastStatusStore()`, it can compute the real refresh interval and then use this one as refresh interval for the binding.
There is absolutely no need for this refresh interval to be defined by the user.

I hope I am clear.

**Comment by lolodomo** [comment]
PR updated: the binding is now searching automatically the right refresh interval for the weather station and the home coach.

**Comment by lolodomo** [comment]
@clinique : can you have a look please ? 
I have tested and it is working well, the binding is finding that the refresh interval to be used is 600 seconds after few minutes.

**Comment by lolodomo** [comment]
An explanation of the algorythm: at startup, the data from the device are requested every 2 minutes until a change of the timestamp is detected. At this time, the difference between the old and the new timestamps gives us the refresh interval for the device which is then used by the binding to schedule a refresh of data.

**Comment by clinique** [comment]
@lolodomo : sorry for delayed answered. Strangely, I was on holidays :) 
I like the idea of auto adapting. Taking a look at your code right now.

**Comment by lolodomo** [comment]
@martinvw : can you merge please ?

File: addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/RefreshStrategy.java [hunk 1]
```diff
@@ -23,17 +26,45 @@
  */
 @NonNullByDefault
 public class RefreshStrategy {
-    private static final int DEFAULT_DELAY = 30;
+
+    private Logger logger = LoggerFactory.getLogger(RefreshStrategy.class);
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/RefreshStrategy.java]: For another time, I like the loggers to be final :-)

File: addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/RefreshStrategy.java [hunk 2]
```diff
@@ -23,17 +26,45 @@
...
         expireData();
     }
 
     public void setDataTimeStamp(Integer dataTimestamp) {
+        if (searchRefreshInterval) {
+            if (dataTimestamp0 == null) {
+                dataTimestamp0 = dataTimestamp;
+                logger.debug("First data timestamp is {}", dataTimestamp0);
+            } else if (dataTimestamp.intValue() > dataTimestamp0.intValue()) {
+                dataValidityPeriod = (dataTimestamp.intValue() - dataTimestamp0.intValue()) * 1000;
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/RefreshStrategy.java]: I see a lot of mapping between MS en S, maybe we could just (I know breaking) consider changing it to seconds.

**Pull request merged**